### PR TITLE
[#164829362] Social login implementation backend facing

### PR DESCRIPTION
**Pull request by kirega**
#### Description
Adds user authentication with social login for Google, Facebook, and Twitter. 
Users who wish to use any one of these social platforms with automatically get access to the site after successfully being authorized by these 3rd party sites. 
#### Type of change
Separates the logic for handling the way that the different social platforms have structured their social logins.  There are three separate endpoints for each social site that only accept a `POST`  method of the following nature:
##### `POST` `localhost/api/users/google/`
- Login using google 
```
{
         "access_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjZmNjc4MWJhNzExOTlhNjU4ZTc2MGFhNWFhOTNlNWZjM2RjNzUyYjUiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJhY2NvdW50cy5nb29nbGUuY29tIiwiYXpwIjoiNDk5MTIxNjI4Mzc2LXU2bjQ1OW4ybGVkNTJzMHE0MWg4aTc1MXFzNjk2N29zLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwiYXVkIjoiNDk5MTIxNjI4Mzc2LXU2bjQ1OW4ybGVkNTJzMHE0MWg4aTc1MXFzNjk2N29zLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwic3ViIjoiMTE1MzAzNTgxMzk0NTEyNjgyNjAwIiwiZW1haWwiOiJqb2VraXJlZ2FAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImF0X2hhc2giOiJRZDJVUmZZZGdValRqTXZmQXFoZXdnIiwibmFtZSI6ImpvZSBraXJlZ2EiLCJwaWN0dXJlIjoiaHR0cHM6Ly9saDUuZ29vZ2xldXNlcmNvbnRlbnQuY29tLy1rcTBGMFZCSm9Tcy9BQUFBQUFBQUFBSS9BQUFBQUFBQUFBQS9BQ0hpM3JlYTFaNm50LXFaTUVqVk5JeW1yRmlBQXdVeVFnL3M5Ni1jL3Bob3RvLmpwZyIsImdpdmVuX25hbWUiOiJqb2UiLCJmYW1pbHlfbmFtZSI6ImtpcmVnYSIsImxvY2FsZSI6ImVuIiwiaWF0IjoxNTU0ODQyOTM3LCJleHAiOjE1NTQ4NDY1MzcsImp0aSI6ImI1MzYxMGEyOTFkNmYzNWYyMzhlYjRjZDNjNzdmNjEzZWMwNjMyOTIifQ.r5yXGPAL04uwsNczquypReGLyP6saHpgXdxNZT2kBUe5zKWditS25VKfXJjmCK6vnz1oJm8W5WItEDq8M0wOK_0gM9i1UMXjzKpCVB4RAYVvSoUu6nKWeV7TUu0BBSFY-MnyM62VoHekzQdNnmqW14lunX6dP6iyMsM6MEV51MYi53VwM96odrZ1e2HVWn_5dwWvef4HkkP8knrsbRp1YFp15odB37RMWQzl2n1WF4Id5BubrlMiA7LvsbdthwPnBhIs11PcOabEqZmSxjk4N0Zss2yEofKsBpiuri5umfBHXrjDwtq9EFuuZWNz6veqjImMYLbPG5B1f-SRyH647Q"
} 
```
On success the payload should return a user access token from out backend in the form of :
```

{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```

- Login using facebook 
##### `POST` `localhost/api/users/facebook/`
```
{
         "access_token": "< access_token>"
} 
```
On success the payload should return a user access token from out backend in the form of :
```
{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```
- Login using Twitter 
##### `POST` `localhost/api/users/twitter/`
This endpoint utilizes a slightly different work flow because it uses Oauth1. 
To make the `POST` send the `access_token_key` and `access_token_secret` separated with a space  a single payload
```
{
         "access_token": "< access_token_key> <access_token_secret >"
} 
```
On success the payload should return a user access token from out backend in the form of :
```
{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```
Incase of invalid tokens the endpoint returns the error in the form of 
```
{
    "errors": {
        "access_token": [
            "Invalid token please try again"
        ]
    }
}
```
#### How Can This Be Tested?
- Clone the repo and navigate to the root directory
- Create a virtual environment by typing `virtualenv venv` and activate it
- Install requirements by typing `pip install -r dev-requirements.txt`
- Setup a database and ensure that your `.env` has the following variables:
```
export SECRET_KEY
export DEBUG
export DATABASE_URL
export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY
export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET
export SOCIAL_AUTH_TWITTER_KEY
export SOCIAL_AUTH_TWITTER_SECRET
export AH_HOST
export CLOUDINARY_NAME
export CLOUDINARY_API_KEY
export CLOUDINARY_API_SECRET
export SENDGRID_API_KEY
export FROM_EMAIL
export URL
export EMAIL_HOST=
export EMAIL_HOST_PASSWORD
export EMAIL_PORT
export EMAIL_USE_TLS
export DOMAIN

```
- Run the tests by typing `python manage.py test`
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164829370

### CODE REVIEW

**Review by Raywire — COMMENTED** [review]
I have tested the google login locally it is working, please update the payload to include more details like username or email the user on login.
Screenshot 2019-04-10 at 12 31 13

**Comment by kirega** [comment]
@Raywire  I appreciate that you took time to review. I considered deeply the pros and cons of having the details of the user alongside the token and I came to the conclusion that it was unnecessary to add the user detail. The reasoning behind this conclusion is that the token is already encoded with user details. I hope this makes sense and that you may consider my approach.

**Comment by kirega** [comment]
@Raywire I appreciate that you took time to review. As I consider to add in more user details. Would you mind elaborating on the benefit of taking such a course of action?

**Review by kihuha — DISMISSED** [review]
Functionality works

**Comment by WNjihia** [comment]
Kindly update the travis.yml to remove the selenium setup

**Review by kihuha — DISMISSED** [review]
Works okay

**Review by Raywire — DISMISSED** [review]
Social auth works well and returns a valid token that can be used to log in to the app.

**Comment by kihuha** [comment]
Quick one, I am trying to access the profiles endpoints and they require a username to be appended to the url. How would I achieve this if I have logged in using any social auth platform? Is there a way of getting the username from the social auth token?

File: requirements.txt [hunk 1]
```diff
@@ -61,3 +61,7 @@ soupsieve==1.9
 typed-ast==1.3.1
 urllib3==1.24.1
 wrapt==1.11.1
+selenium==3.141.0
```
Comment (WNjihia) [inline requirements.txt]: Do we still need to have selenium as a requirement?

**Review by WNjihia — DISMISSED** [review]

File: requirements.txt (hunk 1 above)
Comment (kirega) [reply requirements.txt]: No we do not use this anymore. I will remove it.

**Comment by kirega** [comment]
@kihuha Nice point you have raised there. I had not envisioned that scenario. I can nest the users details alongside the login token to assist you with that.

**Comment by kirega** [comment]
I believe with this new build the username will be incorporated

**Comment by kirega** [comment]
@kihuha You will find the issue fixed in the latest commit

**Review by Kyppy — DISMISSED** [review]
Good job correcting the PR instructions.

**Review by Raywire — DISMISSED** [review]
Social auth works well and returns a valid token that can be used to log in to the app.

**Review by Raywire — APPROVED** [review]
I am satisfied by how this feature works in the backend, tested it locally everything is in order, great job.

**Review by ipaullly — APPROVED** [review]
have run the functionality locally, its an interesting feature and works as you described it.

**Pull request merged**